File: day2/solve.ts
import { max, sum, sumBy } from 'lodash';
import { parseLines, solve } from '../utils/typescript';

type CubeCount = {
  game: number;
  red: number;
  blue: number;
  green: number;
};

function maxCounts(line: string, colour: string) {
  const match = line.match(new RegExp(`\\d+ ${colour}`, 'g'));
  const counts = (match ?? []).map((l) => l.match(/\d+/)).map(Number);
  return max(counts);
}

function countCubes(line: string, i: number): CubeCount {
  return {
    game: i + 1,
    red: maxCounts(line, 'red'),
    blue: maxCounts(line, 'blue'),
    green: maxCounts(line, 'green'),
  };
}

function part1(input: CubeCount[]) {
  const validInputs = input.filter(
    ({ red, blue, green }) => red < 13 && blue < 15 && green < 14,
  );
  return sumBy(validInputs, 'game');
}

function part2(input: CubeCount[]) {
  const scores = input.map(({ red, blue, green }) => red * blue * green);
  return sum(scores);
}

solve({
  part1,
  test1: 8,
  part2,
  test2: 2286,
  parser: (input) => input.split('\n').map(countCubes),
});

File: day2/solve.rb
require 'test/unit'
extend Test::Unit::Assertions

def part1(input)
    input.each_line.with_index.map do |line, i|
        next 0 if line.scan(/\d+ red/).map(&:to_i).max > 12
        next 0 if line.scan(/\d+ green/).map(&:to_i).max > 13
        next 0 if line.scan(/\d+ blue/).map(&:to_i).max > 14
        i + 1
    end.reduce(:+)
end
    

def part2(input)
    input.each_line.map do |line, i|
        ['red', 'green', 'blue'].map do |color|
            line.scan(/\d+ #{color}/).map(&:to_i).max
        end.reduce(:*)
    end.reduce(:+)
end

assert_equal 8, part1(File.read("#{__dir__}/test1.txt"))
assert_equal 2286, part2(File.read("#{__dir__}/test2.txt"))
